File: apps/backend/src/parsers/gstr2b.parser.ts
/**
 * @file apps/backend/src/parsers/gstr2b.parser.ts
 * @description Excel parser for GSTR-2B multi-sheet files.
 * Extracts file header metadata, B2B invoice rows, and ITC summary totals
 * from the exact column and sheet format used in real GSTR-2B downloads.
 *
 * Expected sheets:
 *   - "Read me"            — File header (Financial Year, Tax Period, GSTIN, etc.)
 *   - "B2B"               — B2B invoice rows
 *   - "ITC Available"      — ITC available summary (Part A and Part B)
 *   - "ITC not available"  — Blocked/ineligible ITC summary (Part A and Part B)
 *
 * B2B columns (exact names):
 *   GSTIN of supplier | Trade/Legal name | Invoice number | Invoice type |
 *   Invoice date | Invoice Value(₹) | Place of supply | Supply Attracts Reverse Charge |
 *   Rate(%) | Taxable value | Integrated Tax | Central Tax | State/UT tax | Cess |
 *   GSTR-1/IFF/GSTR-5 Period | GSTR-1/IFF/GSTR-5 Filing Date | ITC Availability |
 *   Reason | Applicable % of Tax Rate | Source | IRN | IRN date
 *
 * Phase 2: Parser implemented based on real data samples.
 * Phase 4: Called during file upload to persist records into MongoDB.
 */

import ExcelJS from 'exceljs';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Metadata extracted from the "Read me" sheet */
export interface Gstr2BMetadata {
  financialYear: string;
  taxPeriod: string;
  buyerGstin: string;
  legalName: string;
  tradeName: string;
  dateOfGeneration: string;
}

/** A single parsed B2B invoice row */
export interface Gstr2BInvoice {
  sheetName: string;
  supplierGstin: string;
  supplierTradeName: string;
  invoiceNumber: string;
  invoiceType: string;
  invoiceDate: string;
  invoiceValue: number;
  placeOfSupply: string;
  supplyAttractsReverseCharge: string;
  taxRate: number | null;
  taxableValue: number;
  integratedTax: number;
  centralTax: number;
  stateUtTax: number;
  cess: number;
  gstr1Period: string | null;
  gstr1FilingDate: string | null;
  itcAvailability: string;
  itcUnavailableReason: string | null;
  applicableTaxRatePercent: number | null;
  source: string;
  irn: string | null;
  irnDate: string | null;
}

/** ITC summary totals from ITC Available / ITC not available sheets */
export interface ItcSummary {
  partA: Record<string, number>;
  partB: Record<string, number>;
}

/** Full result returned by the GSTR-2B parser */
export interface Gstr2BParseResult {
  metadata: Gstr2BMetadata;
  b2bInvoices: Gstr2BInvoice[];
  itcAvailableSummary: ItcSummary;
  itcNotAvailableSummary: ItcSummary;
}

// ---------------------------------------------------------------------------
// Required B2B columns (exact names from real files)
// ---------------------------------------------------------------------------

const B2B_REQUIRED_COLUMNS = [
  'GSTIN of supplier',
  'Trade/Legal name',
  'Invoice number',
  'Invoice type',
  'Invoice date',
  'Invoice Value(₹)',
  'Place of supply',
  'Supply Attracts Reverse Charge',
  'Rate(%)',
  'Taxable value',
  'Integrated Tax',
  'Central Tax',
  'State/UT tax',
  'Cess',
  'GSTR-1/IFF/GSTR-5 Period',
  'GSTR-1/IFF/GSTR-5 Filing Date',
  'ITC Availability',
  'Reason',
  'Applicable % of Tax Rate',
  'Source',
  'IRN',
  'IRN date',
] as const;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Extracts a plain primitive value from an ExcelJS cell value.
 * ExcelJS cells can contain rich text objects, formula results, hyperlinks, etc.
 * This function always returns a string, number, boolean, Date, or null.
 *
 * @param cellValue - Raw cell value from ExcelJS
 */
function extractCellValue(
  cellValue: ExcelJS.CellValue,
): string | number | boolean | Date | null {
  if (cellValue === null || cellValue === undefined) return null;
  if (typeof cellValue === 'string') return cellValue;
  if (typeof cellValue === 'number') return cellValue;
  if (typeof cellValue === 'boolean') return cellValue;
  if (cellValue instanceof Date) return cellValue;

  // Formula cell: return the cached result
  if (typeof cellValue === 'object' && 'result' in cellValue) {
    const result = (cellValue as ExcelJS.CellFormulaValue).result;
    return extractCellValue(result as ExcelJS.CellValue);
  }

  // Rich text cell: concatenate all text runs
  if (typeof cellValue === 'object' && 'richText' in cellValue) {
    return (cellValue as ExcelJS.CellRichTextValue).richText
      .map((run) => run.text)
      .join('');
  }

  // Hyperlink cell: return the display text
  if (typeof cellValue === 'object' && 'text' in cellValue) {
    const text = (cellValue as ExcelJS.CellHyperlinkValue).text;
    return typeof text === 'string' ? text : null;
  }

  return null;
}

/**
 * Safely converts an ExcelJS cell value to a number.
 * Returns 0 for null, undefined, empty string, or non-numeric values.
 *
 * @param cellValue - Raw cell value from ExcelJS
 */
function toNumber(cellValue: ExcelJS.CellValue): number {
  const value = extractCellValue(cellValue);
  if (value === null || value === '') return 0;
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const cleaned = value.replace(/,/g, '').trim();
    const parsed = parseFloat(cleaned);
    return isNaN(parsed) ? 0 : parsed;
  }
  return 0;
}

/**
 * Safely converts an ExcelJS cell value to a number, returning null for empty values.
 * Used for optional numeric fields.
 *
 * @param cellValue - Raw cell value from ExcelJS
 */
function toNumberOrNull(cellValue: ExcelJS.CellValue): number | null {
  const value = extractCellValue(cellValue);
  if (value === null || value === undefined || value === '') return null;
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const cleaned = value.replace(/,/g, '').trim();
    if (cleaned === '') return null;
    const parsed = parseFloat(cleaned);
    return isNaN(parsed) ? null : parsed;
  }
  return null;
}

/**
 * Safely converts an ExcelJS cell value to a string.
 * Returns empty string for null or undefined.
 *
 * @param cellValue - Raw cell value from ExcelJS
 */
function toString(cellValue: ExcelJS.CellValue): string {
  const value = extractCellValue(cellValue);
  if (value === null || value === undefined) return '';
  if (value instanceof Date) return value.toLocaleDateString('en-GB');
  return String(value).trim();
}

/**
 * Converts an ExcelJS cell value to a nullable string.
 * Returns null for empty/missing values.
 *
 * @param cellValue - Raw cell value from ExcelJS
 */
function toStringOrNull(cellValue: ExcelJS.CellValue): string | null {
  const str = toString(cellValue);
  return str === '' ? null : str;
}

/**
 * Reads all rows from an ExcelJS worksheet as a 2D array of raw cell values.
 * Skips completely empty rows.
 *
 * @param worksheet - The ExcelJS worksheet to read
 */
function readAllRows(worksheet: ExcelJS.Worksheet): ExcelJS.CellValue[][] {
  const rows: ExcelJS.CellValue[][] = [];
  worksheet.eachRow({ includeEmpty: false }, (row) => {
    // row.values is 1-indexed (index 0 is null); slice from index 1
    const values = (row.values as ExcelJS.CellValue[]).slice(1);
    rows.push(values);
  });
  return rows;
}

/**
 * Extracts metadata from the "Read me" sheet.
 * Expected layout (each key-value pair in two adjacent columns):
 *   Financial Year:     2022-23
 *   Tax Period:         March
 *   GSTIN:              36BOTPJ1566A1ZX
 *   Legal Name:         JUMANA
 *   Trade Name:         NEXUS PROFILES
 *   Date of generation: 14/04/2023
 *
 * @param rows - All rows from the "Read me" sheet as arrays
 */
function extractReadMeMetadata(rows: ExcelJS.CellValue[][]): Gstr2BMetadata {
  let financialYear = '';
  let taxPeriod = '';
  let buyerGstin = '';
  let legalName = '';
  let tradeName = '';
  let dateOfGeneration = '';

  for (const row of rows) {
    const label = toString(row[0]).toLowerCase();
    const value = toString(row[1]);

    if (label.includes('financial year')) {
      financialYear = value;
    } else if (label.includes('tax period')) {
      taxPeriod = value;
    } else if (label.startsWith('gstin')) {
      buyerGstin = value;
    } else if (label.includes('legal name')) {
      legalName = value;
    } else if (label.includes('trade name')) {
      tradeName = value;
    } else if (label.includes('date of generation')) {
      dateOfGeneration = value;
    }
  }

  return { financialYear, taxPeriod, buyerGstin, legalName, tradeName, dateOfGeneration };
}

/**
 * Parses invoice rows from a B2B-type worksheet.
 * Finds the header row by looking for "GSTIN of supplier" in any cell,
 * then maps each subsequent row using the confirmed column names.
 *
 * @param worksheet  - The ExcelJS worksheet object
 * @param sheetLabel - Sheet name used for logging and stored in each record
 */
function parseB2BSheet(worksheet: ExcelJS.Worksheet, sheetLabel: string): Gstr2BInvoice[] {
  const allRows = readAllRows(worksheet);

  if (allRows.length === 0) {
    console.log(`[GSTR-2B Parser] Sheet "${sheetLabel}" is empty`);
    return [];
  }

  // Find the header row (contains "GSTIN of supplier")
  let headerRowIndex = -1;
  for (let i = 0; i < allRows.length; i++) {
    const rowStrings = allRows[i].map((cell) => toString(cell));
    if (rowStrings.some((cell) => cell.toLowerCase().includes('gstin of supplier'))) {
      headerRowIndex = i;
      break;
    }
  }

  if (headerRowIndex === -1) {
    console.log(`[GSTR-2B Parser] No header row found in sheet "${sheetLabel}" — skipping`);
    return [];
  }

  const headerRow = allRows[headerRowIndex].map((cell) => toString(cell));

  // Verify all required columns are present
  const missingColumns = B2B_REQUIRED_COLUMNS.filter((col) => !headerRow.includes(col));
  if (missingColumns.length > 0) {
    throw new Error(
      `[GSTR-2B Parser] Sheet "${sheetLabel}" is missing required columns: ${missingColumns.join(', ')}`,
    );
  }

  // Build column index map
  const columnIndex: Record<string, number> = {};
  for (const col of B2B_REQUIRED_COLUMNS) {
    columnIndex[col] = headerRow.indexOf(col);
  }

  const invoices: Gstr2BInvoice[] = [];

  for (let i = headerRowIndex + 1; i < allRows.length; i++) {
    const row = allRows[i];

    // Skip empty rows
    if (!row || row.every((cell) => cell === null || cell === undefined || cell === '')) continue;

    const supplierGstin = toString(row[columnIndex['GSTIN of supplier']]);

    // Skip rows that are section headers or sub-totals (no GSTIN)
    if (!supplierGstin) continue;

    const invoice: Gstr2BInvoice = {
      sheetName: sheetLabel,
      supplierGstin,
      supplierTradeName: toString(row[columnIndex['Trade/Legal name']]),
      invoiceNumber: toString(row[columnIndex['Invoice number']]),
      invoiceType: toString(row[columnIndex['Invoice type']]),
      invoiceDate: toString(row[columnIndex['Invoice date']]),
      invoiceValue: toNumber(row[columnIndex['Invoice Value(₹)']]),
      placeOfSupply: toString(row[columnIndex['Place of supply']]),
      supplyAttractsReverseCharge: toString(row[columnIndex['Supply Attracts Reverse Charge']]),
      taxRate: toNumberOrNull(row[columnIndex['Rate(%)']]),
      taxableValue: toNumber(row[columnIndex['Taxable value']]),
      integratedTax: toNumber(row[columnIndex['Integrated Tax']]),
      centralTax: toNumber(row[columnIndex['Central Tax']]),
      stateUtTax: toNumber(row[columnIndex['State/UT tax']]),
      cess: toNumber(row[columnIndex['Cess']]),
      gstr1Period: toStringOrNull(row[columnIndex['GSTR-1/IFF/GSTR-5 Period']]),
      gstr1FilingDate: toStringOrNull(row[columnIndex['GSTR-1/IFF/GSTR-5 Filing Date']]),
      itcAvailability: toString(row[columnIndex['ITC Availability']]),
      itcUnavailableReason: toStringOrNull(row[columnIndex['Reason']]),
      applicableTaxRatePercent: toNumberOrNull(row[columnIndex['Applicable % of Tax Rate']]),
      source: toString(row[columnIndex['Source']]),
      irn: toStringOrNull(row[columnIndex['IRN']]),
      irnDate: toStringOrNull(row[columnIndex['IRN date']]),
    };

    invoices.push(invoice);
  }

  return invoices;
}

/**
 * Parses an ITC summary sheet (e.g. "ITC Available" or "ITC not available").
 * Extracts key-value totals from Part A and Part B sections.
 *
 * @param worksheet  - The ExcelJS worksheet object
 * @param sheetLabel - Sheet name used for logging
 */
function parseItcSummarySheet(worksheet: ExcelJS.Worksheet, sheetLabel: string): ItcSummary {
  const allRows = readAllRows(worksheet);
  const partA: Record<string, number> = {};
  const partB: Record<string, number> = {};
  let currentPart = '';

  for (const row of allRows) {
    const cellZero = toString(row[0]).toLowerCase();
    const cellOne = toString(row[1]).toLowerCase();

    // Detect part headings
    if (cellZero.includes('part a') || cellOne.includes('part a')) {
      currentPart = 'A';
      continue;
    }
    if (cellZero.includes('part b') || cellOne.includes('part b')) {
      currentPart = 'B';
      continue;
    }

    // Extract label-value pairs
    const label = toString(row[0]);
    // Look for a numeric value in the last non-null column
    const numericValue = [...row]
      .reverse()
      .find((cell) => cell !== null && cell !== undefined && cell !== '');
    const amount = toNumber(numericValue ?? null);

    if (label && amount !== 0) {
      if (currentPart === 'A') {
        partA[label] = amount;
      } else if (currentPart === 'B') {
        partB[label] = amount;
      }
    }
  }

  console.log(
    `[GSTR-2B Parser] ITC summary "${sheetLabel}": Part A keys=${Object.keys(partA).length}, Part B keys=${Object.keys(partB).length}`,
  );

  return { partA, partB };
}

// ---------------------------------------------------------------------------
// Main parser
// ---------------------------------------------------------------------------

/**
 * Parses a GSTR-2B Excel file (multi-sheet format) and returns file metadata,
 * B2B invoice rows, and ITC summary totals.
 *
 * @param fileBuffer - Binary content of the uploaded Excel file
 * @param fileName   - Original file name (used for logging only)
 * @throws Error if required sheets or columns are missing
 */
export async function parseGstr2B(
  fileBuffer: Buffer,
  fileName: string,
): Promise<Gstr2BParseResult> {
  console.log(`[GSTR-2B Parser] Parsing file: ${fileName}`);

  const workbook = new ExcelJS.Workbook();
  // Convert Node.js Buffer to ArrayBuffer to satisfy ExcelJS type definitions
  const arrayBuffer = fileBuffer.buffer.slice(
    fileBuffer.byteOffset,
    fileBuffer.byteOffset + fileBuffer.byteLength,
  ) as ArrayBuffer;
  await workbook.xlsx.load(arrayBuffer);

  const sheetNames = workbook.worksheets.map((ws) => ws.name);
  console.log(`[GSTR-2B Parser] Sheets found: ${sheetNames.join(', ')}`);

  // Find the "Read me" sheet (case-insensitive)
  const readMeSheet = workbook.worksheets.find((ws) =>
    ws.name.toLowerCase().includes('read me'),
  );
  if (!readMeSheet) {
    throw new Error('[GSTR-2B Parser] Required "Read me" sheet not found');
  }

  // Extract metadata from "Read me" sheet
  const readMeRows = readAllRows(readMeSheet);
  const metadata = extractReadMeMetadata(readMeRows);
  console.log(
    `[GSTR-2B Parser] Metadata — FY: ${metadata.financialYear}, Period: ${metadata.taxPeriod}, GSTIN: ${metadata.buyerGstin}`,
  );

  // Find and parse the "B2B" sheet (exact match, case-insensitive)
  const b2bSheet = workbook.worksheets.find((ws) => ws.name.toUpperCase() === 'B2B');
  if (!b2bSheet) {
    throw new Error('[GSTR-2B Parser] Required "B2B" sheet not found');
  }

  const b2bInvoices = parseB2BSheet(b2bSheet, b2bSheet.name);
  console.log(`[GSTR-2B Parser] Parsed ${b2bInvoices.length} B2B invoice rows`);

  // Parse "ITC Available" sheet (optional — warn if missing)
  let itcAvailableSummary: ItcSummary = { partA: {}, partB: {} };
  const itcAvailSheet = workbook.worksheets.find(
    (ws) =>
      ws.name.toLowerCase().includes('itc available') &&
      !ws.name.toLowerCase().includes('not'),
  );
  if (itcAvailSheet) {
    itcAvailableSummary = parseItcSummarySheet(itcAvailSheet, itcAvailSheet.name);
  } else {
    console.warn('[GSTR-2B Parser] "ITC Available" sheet not found — summary will be empty');
  }

  // Parse "ITC not available" sheet (optional — warn if missing)
  let itcNotAvailableSummary: ItcSummary = { partA: {}, partB: {} };
  const itcNotAvailSheet = workbook.worksheets.find((ws) =>
    ws.name.toLowerCase().includes('itc not available'),
  );
  if (itcNotAvailSheet) {
    itcNotAvailableSummary = parseItcSummarySheet(itcNotAvailSheet, itcNotAvailSheet.name);
  } else {
    console.warn(
      '[GSTR-2B Parser] "ITC not available" sheet not found — summary will be empty',
    );
  }

  return { metadata, b2bInvoices, itcAvailableSummary, itcNotAvailableSummary };
}

File: apps/backend/app/parsers/gstr2b_parser.py
"""
GSTR-2B Excel parser using openpyxl.
Rewrite of gstr2b.parser.ts with identical logic.
"""

import re
from io import BytesIO
from typing import Optional
import openpyxl
from pydantic import BaseModel


class Gstr2BMetadata(BaseModel):
    financial_year: str = ""
    tax_period: str = ""
    buyer_gstin: str = ""
    legal_name: str = ""
    trade_name: str = ""
    date_of_generation: str = ""


class Gstr2BInvoice(BaseModel):
    sheet_name: str = ""
    supplier_gstin: str = ""
    supplier_trade_name: str = ""
    invoice_number: str = ""
    invoice_type: str = ""
    invoice_date: str = ""
    invoice_value: float = 0.0
    place_of_supply: str = ""
    supply_attracts_reverse_charge: str = ""
    tax_rate: Optional[float] = None
    taxable_value: float = 0.0
    integrated_tax: float = 0.0
    central_tax: float = 0.0
    state_ut_tax: float = 0.0
    cess: float = 0.0
    gstr1_period: Optional[str] = None
    gstr1_filing_date: Optional[str] = None
    itc_availability: str = ""
    itc_unavailable_reason: Optional[str] = None
    applicable_tax_rate_percent: Optional[float] = None
    source: str = ""
    irn: Optional[str] = None
    irn_date: Optional[str] = None


class ItcSummary(BaseModel):
    part_a: dict = {}
    part_b: dict = {}


class Gstr2BParseResult(BaseModel):
    metadata: Gstr2BMetadata
    b2b_invoices: list[Gstr2BInvoice]
    itc_available_summary: ItcSummary
    itc_not_available_summary: ItcSummary


CORE_COLUMNS = ("party_gstin", "invoice_number", "invoice_date", "taxable_amount")

GST_HEADER_KEYWORDS = ("gstin", "tax", "invoice", "amount", "date", "particulars", "supplier")

COLUMN_ALIASES: dict[str, list[str]] = {
    "party_gstin": ["GSTIN of supplier", "Supplier GSTIN", "GSTIN/UIN"],
    "supplier_name": ["Trade/Legal name", "Supplier Trade Name", "Particulars"],
    "invoice_number": ["Invoice number", "Invoice No.", "Vch No."],
    "invoice_type": ["Invoice type", "Vch Type"],
    "invoice_date": ["Invoice date", "Invoice Date", "Date"],
    "invoice_value": ["Invoice Value(₹)", "Invoice Value", "Invoice Amount"],
    "place_of_supply": ["Place of supply"],
    "reverse_charge": ["Supply Attracts Reverse Charge", "Reverse Charge"],
    "tax_rate": ["Rate(%)", "Tax Rate"],
    "taxable_amount": ["Taxable value", "Taxable Value", "Taxable Amount", "Taxable Income"],
    "igst": ["Integrated Tax", "IGST"],
    "cgst": ["Central Tax", "CGST"],
    "sgst": ["State/UT tax", "State/UT Tax", "State Tax", "SGST", "SGST/UTGST"],
    "cess": ["Cess", "Cess Amount"],
    "gstr1_period": ["GSTR-1/IFF/GSTR-5 Period"],
    "gstr1_filing_date": ["GSTR-1/IFF/GSTR-5 Filing Date"],
    "itc_availability": ["ITC Availability"],
    "reason": ["Reason"],
    "applicable_tax_rate_percent": ["Applicable % of Tax Rate"],
    "source": ["Source"],
    "irn": ["IRN"],
    "irn_date": ["IRN date"],
}


def safe_number(val) -> float:
    if val is None or val == "":
        return 0.0
    try:
        if isinstance(val, str):
            cleaned = val.replace("₹", "").replace(",", "").strip()
            return float(cleaned) if cleaned else 0.0
        return float(val)
    except (ValueError, TypeError):
        return 0.0


def safe_optional_number(val) -> Optional[float]:
    if val is None or val == "":
        return None
    try:
        return float(val)
    except (ValueError, TypeError):
        return None


def safe_str(val) -> str:
    if val is None:
        return ""
    return str(val).strip()


def safe_optional_str(val) -> Optional[str]:
    if val is None or safe_str(val) == "":
        return None
    return str(val).strip()


def _normalize_header(value) -> str:
    text = safe_str(value).replace("\n", " ").replace("\r", " ")
    return re.sub(r"[^a-z0-9]", "", text.lower())


def _detect_header_row(all_rows: list[tuple]) -> int:
    best_idx = -1
    best_score = -1
    for i, row in enumerate(all_rows[:15]):
        if not row:
            continue
        score = 0
        for cell in row:
            normalized = _normalize_header(cell)
            if not normalized:
                continue
            if any(keyword in normalized for keyword in GST_HEADER_KEYWORDS):
                score += 1
        if score > best_score:
            best_score = score
            best_idx = i
    return best_idx if best_score > 0 else -1


def _build_column_map(header_row: tuple) -> dict[str, int]:
    alias_lookup: dict[str, str] = {}
    for canonical, aliases in COLUMN_ALIASES.items():
        for alias in aliases:
            alias_lookup[_normalize_header(alias)] = canonical

    resolved: dict[str, int] = {}
    for idx, cell in enumerate(header_row):
        normalized = _normalize_header(cell)
        if not normalized:
            continue
        canonical = alias_lookup.get(normalized)
        if canonical and canonical not in resolved:
            resolved[canonical] = idx
    return resolved


def _extract_readme_metadata(ws) -> Gstr2BMetadata:
    """Extract metadata from 'Read me' sheet."""
    metadata = Gstr2BMetadata()
    for row in ws.iter_rows(values_only=True):
        if not row or row[0] is None:
            continue
        key = safe_str(row[0]).lower()
        value = safe_str(row[1]) if len(row) > 1 and row[1] is not None else ""
        if "financial year" in key:
            metadata.financial_year = value
        elif "tax period" in key:
            metadata.tax_period = value
        elif "gstin" in key:
            metadata.buyer_gstin = value
        elif "legal name" in key:
            metadata.legal_name = value
        elif "trade name" in key:
            metadata.trade_name = value
        elif "date of generation" in key:
            metadata.date_of_generation = value
    return metadata


def _parse_b2b_sheet(ws) -> list[Gstr2BInvoice]:
    """Parse the B2B sheet into invoice records."""
    all_rows = list(ws.iter_rows(values_only=True))
    header_row_idx = _detect_header_row(all_rows)

    if header_row_idx == -1:
        return []
    col_map = _build_column_map(all_rows[header_row_idx])
    if any(column not in col_map for column in CORE_COLUMNS):
        return []

    invoices: list[Gstr2BInvoice] = []
    for row in all_rows[header_row_idx + 1:]:
        if not row or all(cell is None for cell in row):
            continue

        def get_cell(col_name: str):
            idx = col_map.get(col_name)
            if idx is None or idx >= len(row):
                return None
            return row[idx]

        supplier_gstin = safe_str(get_cell("party_gstin")).upper().replace(" ", "")
        invoice_number = safe_str(get_cell("invoice_number"))
        invoice_date = safe_str(get_cell("invoice_date"))
        if not supplier_gstin or not invoice_number or not invoice_date:
            continue

        taxable_value = safe_number(get_cell("taxable_amount"))
        integrated_tax = safe_number(get_cell("igst"))
        central_tax = safe_number(get_cell("cgst"))
        state_ut_tax = safe_number(get_cell("sgst"))
        cess = safe_number(get_cell("cess"))
        invoice_value = safe_number(get_cell("invoice_value"))
        if invoice_value == 0.0:
            invoice_value = round(taxable_value + integrated_tax + central_tax + state_ut_tax + cess, 2)

        invoice = Gstr2BInvoice(
            sheet_name="B2B",
            supplier_gstin=supplier_gstin,
            supplier_trade_name=safe_str(get_cell("supplier_name")),
            invoice_number=invoice_number,
            invoice_type=safe_str(get_cell("invoice_type")),
            invoice_date=invoice_date,
            invoice_value=invoice_value,
            place_of_supply=safe_str(get_cell("place_of_supply")),
            supply_attracts_reverse_charge=safe_str(get_cell("reverse_charge")),
            tax_rate=safe_optional_number(get_cell("tax_rate")),
            taxable_value=taxable_value,
            integrated_tax=integrated_tax,
            central_tax=central_tax,
            state_ut_tax=state_ut_tax,
            cess=cess,
            gstr1_period=safe_optional_str(get_cell("gstr1_period")),
            gstr1_filing_date=safe_optional_str(get_cell("gstr1_filing_date")),
            itc_availability=safe_str(get_cell("itc_availability")),
            itc_unavailable_reason=safe_optional_str(get_cell("reason")),
            applicable_tax_rate_percent=safe_optional_number(get_cell("applicable_tax_rate_percent")),
            source=safe_str(get_cell("source")),
            irn=safe_optional_str(get_cell("irn")),
            irn_date=safe_optional_str(get_cell("irn_date")),
        )
        invoices.append(invoice)

    return invoices


def _parse_itc_summary_sheet(ws) -> ItcSummary:
    """Parse ITC summary sheet into ItcSummary."""
    part_a: dict = {}
    part_b: dict = {}
    current_part = None
    for row in ws.iter_rows(values_only=True):
        if not row or row[0] is None:
            continue
        cell = safe_str(row[0]).lower()
        if "part a" in cell:
            current_part = "a"
        elif "part b" in cell:
            current_part = "b"
        elif current_part and row[0] is not None:
            key = safe_str(row[0])
            value = safe_number(row[1]) if len(row) > 1 else 0.0
            if current_part == "a":
                part_a[key] = value
            else:
                part_b[key] = value
    return ItcSummary(part_a=part_a, part_b=part_b)


def parse_gstr2b(file_bytes: bytes, file_name: str) -> Gstr2BParseResult:
    """Parse GSTR-2B Excel file and return structured data."""
    wb = openpyxl.load_workbook(BytesIO(file_bytes), read_only=True, data_only=True)
    try:
        # Extract metadata from "Read me" sheet
        metadata = Gstr2BMetadata()
        for sheet_name in wb.sheetnames:
            if sheet_name.lower().strip() == "read me":
                metadata = _extract_readme_metadata(wb[sheet_name])
                break

        # Parse B2B invoices
        b2b_invoices: list[Gstr2BInvoice] = []
        for sheet_name in wb.sheetnames:
            if sheet_name.strip().upper() == "B2B":
                b2b_invoices = _parse_b2b_sheet(wb[sheet_name])
                break
        if not b2b_invoices and wb.sheetnames:
            # Graceful fallback for ERP exports where sheet is not literally named "B2B".
            b2b_invoices = _parse_b2b_sheet(wb[wb.sheetnames[0]])

        # Parse ITC summary sheets
        itc_available_summary = ItcSummary()
        itc_not_available_summary = ItcSummary()
        for sheet_name in wb.sheetnames:
            name_lower = sheet_name.lower().strip()
            if "itc available" in name_lower and "not" not in name_lower:
                itc_available_summary = _parse_itc_summary_sheet(wb[sheet_name])
            elif "itc not available" in name_lower:
                itc_not_available_summary = _parse_itc_summary_sheet(wb[sheet_name])

        return Gstr2BParseResult(
            metadata=metadata,
            b2b_invoices=b2b_invoices,
            itc_available_summary=itc_available_summary,
            itc_not_available_summary=itc_not_available_summary,
        )
    finally:
        wb.close()
